shared/ui/textField/TextField.tsx

```diff
@@ -1,102 +1,123 @@
-import { forwardRef, useState } from 'react';
+import { ChangeEvent, ForwardedRef, ReactNode, forwardRef, useState } from 'react';
 
 import Close from '@/assets/icons/close.svg?url';
 import Eye from '@/assets/icons/eye-outline.svg?url';
 import SearchOutline from '@/assets/icons/searchOutline.svg?url';
 import { Typography } from '@/shared/ui/typography';
+import { clsx } from 'clsx';
 import Image from 'next/image';
 
 import s from 'shared/ui/textField/TextField.module.scss';
 
-// type TextFieldType = 'input' | 'textarea';
+type SharedProps = {
+  className?: string;
+  disabled?: boolean;
+  error?: string;
+  label?: string;
+  onBlur?: () => void;
+  onChange: (value: string) => void;
+  onFocus?: () => void;
+  placeholder?: string;
+  required?: boolean;
+  value: string;
+};
 
+// If you pass this type as a generic parameter, TS will prevent you from using props specific to union members.
+// To be able to pass these props, you can extract the desired union member using the Extract utility type.
+// https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union
 export type TextFieldProps =
-  // <T extends TextFieldType = 'input'>
+  | (SharedProps & {
+      as?: 'input';
+      inputType?: 'password' | 'search' | 'text';
+    })
+  | (SharedProps & { as: 'textarea' });
 
-  {
-    as?: 'input' | 'textarea';
-    className?: string;
-    disabled?: boolean;
-    errors?: string;
-    inputtype?: 'password' | 'text';
-    isSearchInput?: boolean;
-    label?: string;
-    onChange: (e: string) => void;
-    onFocus?: () => void;
-    placeholder?: string;
-    required?: boolean;
-    value: string;
-  };
+// Forces the 'ref' type match the 'as' prop type. You should not specify the 'as' props dynamically.
+// Use only the literal value to avoid possible errors with non-existent properties of the target element pointed by the 'ref'.
+// prettier-ignore
+export type TextFieldComponent = {
+  (props: Extract<TextFieldProps, { as: 'textarea'}> & { ref?: ForwardedRef<HTMLTextAreaElement> }): ReactNode;
+  (props: Extract<TextFieldProps, { as?: 'input'}> & { ref?: ForwardedRef<HTMLInputElement> }): ReactNode;
+}
```

Comment:
This is an implementation of a feature called [function overloading](https://blog.logrocket.com/implementing-function-overloading-typescript/).

The purpose of the `TextFieldComponent` type is to force the component consumer to pass a strictly typed `ref` type corresponding to its derived `props` types, depending on the value of the `as` parameter, be it `input` or `textarea`.

Pass `input` as `as` prop value or don't pass value at all to tell Typescript that this is an input-like component.
Pass `textarea` as `as` prop value  to tell Typescript that this is a textarea-like component.

The following error may occur if you pass `ref` of the wrong type. In this case you need to either type `ref` correctly or change the value of the `as` property:

![ref-error-example](https://github.com/Iplatos/inctagram/assets/93154182/24a66d71-700c-418e-a880-67e6f2b60bb7)